src/partition_chunk.c

```diff
@@ -0,0 +1,338 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+#include <postgres.h>
+#include "access/attmap.h"
+#include "access/toast_compression.h"
+#include <catalog/heap.h>
+#include <commands/tablecmds.h>
+#include <nodes/makefuncs.h>
+#include <nodes/parsenodes.h>
+#include <rewrite/rewriteManip.h>
+#include "utils/partcache.h"
+
+#include "chunk.h"
+#include "guc.h"
+#include "hypercube.h"
+#include "hypertable.h"
+#include "partition_chunk.h"
+
+/*
+ * Cache and the memory context to store recently created chunks to be attached
+ * as partitions.
+ */
+static HTAB *PartChunkCache = NULL;
+static MemoryContext PartChunkCacheCtx = NULL;
+
+/*
+ * Insert a chunk into the partition cache.
+ */
+void
+ts_partition_cache_insert_chunk(const Hypertable *ht, Oid chunk_relid)
+{
+	PartChunkCacheEntry *entry;
+	bool found;
+
+	if (PartChunkCache == NULL)
+	{
+		if (PartChunkCacheCtx == NULL)
+			PartChunkCacheCtx = AllocSetContextCreate(CacheMemoryContext,
+													  "partition chunk cache",
+													  ALLOCSET_DEFAULT_SIZES);
+
+		HASHCTL ctl;
+		ctl.keysize   = sizeof(Oid);
+		ctl.entrysize = sizeof(PartChunkCacheEntry);
+		ctl.hcxt      = PartChunkCacheCtx;
+
+		PartChunkCache = hash_create("partition chunk cache",
+									 256, /* start small, grows automatically */
+									 &ctl,
+									 HASH_ELEM | HASH_BLOBS | HASH_CONTEXT);
+	}
+
+	MemoryContext oldctx = MemoryContextSwitchTo(PartChunkCacheCtx);
```

Comment:
I don't think this switch into memory context is needed. The `hash_search` call will allocate entries on its internal context, which is a child of the memory context specified in `ctl.hcxt`

```suggestion
```